File: protocol/src/frame.rs
//! Binary-frame stream tagging.
//!
//! ## Transport rule (ADR 0001)
//! - **Text frames** carry a JSON control-plane envelope (`ClientMessage` /
//!   `ServerMessage`).
//! - **Binary frames** carry raw stream bytes, prefixed with a 1-byte stream
//!   tag, so the wire format can grow new streams (files, images) later
//!   without a protocol redesign.
//!
//! Bulk terminal output rides binary frames (zero-copy, no base64 tax). Small
//! or out-of-band payloads use the `TerminalOutput` text message instead.

/// Stream-tag constants: the first byte of every binary frame.
pub mod stream_tag {
    /// Terminal stdout/stderr (multiplexed by the PTY into one stream).
    pub const TERMINAL: u8 = 0x01;
}

/// Wrap raw stream bytes with their 1-byte tag: `[tag][payload...]`.
pub fn encode(tag: u8, payload: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(1 + payload.len());
    out.push(tag);
    out.extend_from_slice(payload);
    out
}

/// Split a tagged binary frame into `(stream_tag, payload)`.
///
/// Returns `None` if the frame is empty (no tag byte).
pub fn decode(frame: &[u8]) -> Option<(u8, &[u8])> {
    let (&tag, rest) = frame.split_first()?;
    Some((tag, rest))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip() {
        let payload = b"hello world";
        let encoded = encode(stream_tag::TERMINAL, payload);
        let (tag, decoded) = decode(&encoded).unwrap();
        assert_eq!(tag, stream_tag::TERMINAL);
        assert_eq!(decoded, payload);
    }

    #[test]
    fn empty_frame_is_none() {
        assert!(decode(&[]).is_none());
    }

    #[test]
    fn tag_only_frame_has_empty_payload() {
        let encoded = encode(stream_tag::TERMINAL, b"");
        let (tag, decoded) = decode(&encoded).unwrap();
        assert_eq!(tag, stream_tag::TERMINAL);
        assert!(decoded.is_empty());
    }
}

File: mobile/lib/protocol/frame.dart
/// Binary-frame tagging, Dart mirror of Rust `protocol::frame`.
///
/// Binary frames are `[1-byte tag][payload bytes...]`. Currently only
/// [terminalTag] (0x01 = PTY stdout) is defined; the switch is written so new
/// streams (files, images) can be added without changing the wire format.

library;

import 'dart:typed_data';

/// First byte of a binary frame: identifies the stream.
const int terminalTag = 0x01;

/// A decoded binary frame.
class Frame {
  final int tag;
  final Uint8List payload;
  const Frame({required this.tag, required this.payload});
}

/// Decode a binary frame into [Frame], or `null` if the frame is empty.
Frame? decodeFrame(Uint8List frame) {
  if (frame.isEmpty) return null;
  return Frame(tag: frame[0], payload: Uint8List.sublistView(frame, 1));
}
